File: packages/node/src/session.ts
import { createCipheriv, createDecipheriv, createHash, randomBytes } from 'crypto';

import { MythosConfigError } from './errors';
import type { MythosSession } from './types';

const ALGORITHM = 'aes-256-gcm';
const IV_LENGTH = 12;
const AUTH_TAG_LENGTH = 16;

export interface StoredSession extends MythosSession {
  expiresAt: string;
}

function getSessionKey(): Buffer {
  const secret = process.env.MYTHOS_SESSION_SECRET;
  if (!secret) {
    throw new MythosConfigError('MYTHOS_SESSION_SECRET env var is required to encode/decode a Mythos session');
  }
  return createHash('sha256').update(secret).digest();
}

/**
 * Encrypts a MythosSession (including its LLM identity token) into a compact, tamper-proof
 * string sized to fit an HttpOnly cookie. This is what lets a Producer app support any
 * number of its own routes off a single launch/consume, instead of re-verifying the launch
 * token -- which is single-use -- on every page. The Producer sets the returned string as
 * its own cookie (recommended: HttpOnly; Secure; SameSite=Lax); decodeSession() reads it
 * back on any later request with no call to Mythos at all.
 *
 * NOT cross-language compatible with the Python SDK's encode_session/decode_session, even
 * with the same MYTHOS_SESSION_SECRET: this writes iv(12) || authTag(16) || ciphertext,
 * while Python's AESGCM.encrypt() produces ciphertext || tag(16), giving a different byte
 * layout under the same field names. A Producer app must encode and decode its session
 * cookie with the same language's SDK throughout -- there is no supported mixed-language
 * deployment for a single cookie.
 */
export function encodeSession(session: MythosSession): string {
  const key = getSessionKey();
  const iv = randomBytes(IV_LENGTH);
  const cipher = createCipheriv(ALGORITHM, key, iv);
  const plaintext = Buffer.from(JSON.stringify(session), 'utf8');
  const ciphertext = Buffer.concat([cipher.update(plaintext), cipher.final()]);
  const authTag = cipher.getAuthTag();
  return Buffer.concat([iv, authTag, ciphertext]).toString('base64url');
}

/**
 * Decrypts a value produced by encodeSession(). Returns null for anything that isn't a
 * currently-valid session -- tampered, malformed, or past its own llmIdentityExpiresAt --
 * rather than throwing, since "no valid session" is an expected, ordinary outcome for a
 * caller (e.g. redirect back through launch again), not an error condition. A missing
 * MYTHOS_SESSION_SECRET is a real misconfiguration and still throws MythosConfigError.
 */
export function decodeSession(token: string): MythosSession | null {
  const key = getSessionKey();
  try {
    const raw = Buffer.from(token, 'base64url');
    const iv = raw.subarray(0, IV_LENGTH);
    const authTag = raw.subarray(IV_LENGTH, IV_LENGTH + AUTH_TAG_LENGTH);
    const ciphertext = raw.subarray(IV_LENGTH + AUTH_TAG_LENGTH);
    const decipher = createDecipheriv(ALGORITHM, key, iv);
    decipher.setAuthTag(authTag);
    const plaintext = Buffer.concat([decipher.update(ciphertext), decipher.final()]);
    const session = JSON.parse(plaintext.toString('utf8')) as MythosSession;

    if (session.llmIdentityExpiresAt) {
      const expiresAtMs = new Date(session.llmIdentityExpiresAt).getTime();
      // A malformed date string parses to NaN, and NaN <= anything is always false -- so
      // this must be checked explicitly, otherwise a malformed expiry fails open (treated
      // as "not expired") instead of closed like Python's equivalent check.
      if (Number.isNaN(expiresAtMs) || expiresAtMs <= Date.now()) {
        return null;
      }
    }
    return session;
  } catch {
    return null;
  }
}

export function sealSession(session: StoredSession): string {
  const key = getSessionKey();
  const iv = randomBytes(IV_LENGTH);
  const cipher = createCipheriv(ALGORITHM, key, iv);
  const plaintext = Buffer.from(JSON.stringify(session), 'utf8');
  const ciphertext = Buffer.concat([cipher.update(plaintext), cipher.final()]);
  const authTag = cipher.getAuthTag();
  return Buffer.concat([iv, authTag, ciphertext]).toString('base64url');
}

export function openSession(token: string): StoredSession | null {
  const key = getSessionKey();
  try {
    const raw = Buffer.from(token, 'base64url');
    const iv = raw.subarray(0, IV_LENGTH);
    const authTag = raw.subarray(IV_LENGTH, IV_LENGTH + AUTH_TAG_LENGTH);
    const ciphertext = raw.subarray(IV_LENGTH + AUTH_TAG_LENGTH);
    const decipher = createDecipheriv(ALGORITHM, key, iv);
    decipher.setAuthTag(authTag);
    const plaintext = Buffer.concat([decipher.update(ciphertext), decipher.final()]);
    const session: unknown = JSON.parse(plaintext.toString('utf8'));
    if (typeof session !== 'object' || session === null || Array.isArray(session)) return null;
    const stored = session as StoredSession;
    const expiresAtMs = typeof stored.expiresAt === 'string' ? Date.parse(stored.expiresAt) : Number.NaN;
    if (Number.isNaN(expiresAtMs) || expiresAtMs <= Date.now()) return null;
    return stored;
  } catch {
    return null;
  }
}

File: packages/python/mythos_sdk/session.py
import base64
import hashlib
import json
import os
from dataclasses import asdict
from datetime import datetime, timezone

from cryptography.hazmat.primitives.ciphers.aead import AESGCM

from .errors import MythosConfigError
from .types import MythosSession

NONCE_LENGTH = 12


def _get_session_key() -> bytes:
    secret = os.environ.get("MYTHOS_SESSION_SECRET")
    if not secret:
        raise MythosConfigError("MYTHOS_SESSION_SECRET env var is required to encode/decode a Mythos session")
    return hashlib.sha256(secret.encode("utf-8")).digest()


def encode_session(session: MythosSession) -> str:
    """Encrypts a MythosSession (including its LLM identity token) into a compact,
    tamper-proof string sized to fit an HttpOnly cookie. This is what lets a Producer app
    support any number of its own routes off a single launch/consume, instead of
    re-verifying the launch token -- which is single-use -- on every page. The Producer
    sets the returned string as its own cookie (recommended: HttpOnly; Secure;
    SameSite=Lax); decode_session() reads it back on any later request with no call to
    Mythos at all.

    NOT cross-language compatible with the Node SDK's encodeSession/decodeSession, even
    with the same MYTHOS_SESSION_SECRET: AESGCM.encrypt() here produces
    ciphertext || tag(16), appended after the nonce, while Node writes
    iv(12) || authTag(16) || ciphertext -- a different byte layout under the same field
    names. A Producer app must encode and decode its session cookie with the same
    language's SDK throughout -- there is no supported mixed-language deployment for a
    single cookie.
    """
    key = _get_session_key()
    nonce = os.urandom(NONCE_LENGTH)
    plaintext = json.dumps(asdict(session)).encode("utf-8")
    ciphertext = AESGCM(key).encrypt(nonce, plaintext, None)
    return base64.urlsafe_b64encode(nonce + ciphertext).decode("ascii").rstrip("=")


def decode_session(token: str) -> MythosSession | None:
    """Decrypts a value produced by encode_session(). Returns None for anything that isn't
    a currently-valid session -- tampered, malformed, or past its own
    llmIdentityExpiresAt -- rather than raising, since "no valid session" is an expected,
    ordinary outcome for a caller (e.g. redirect back through launch again), not an error
    condition. A missing MYTHOS_SESSION_SECRET is a real misconfiguration and still raises
    MythosConfigError.
    """
    key = _get_session_key()
    try:
        padded = token + "=" * (-len(token) % 4)
        raw = base64.urlsafe_b64decode(padded.encode("ascii"))
        nonce, ciphertext = raw[:NONCE_LENGTH], raw[NONCE_LENGTH:]
        plaintext = AESGCM(key).decrypt(nonce, ciphertext, None)
        data = json.loads(plaintext.decode("utf-8"))

        expires_at = data.get("llmIdentityExpiresAt")
        if expires_at and datetime.fromisoformat(expires_at) <= datetime.now(timezone.utc):
            return None
        return MythosSession(**data)
    except Exception:
        return None


def seal_session(session: MythosSession, expires_at: str) -> str:
    key = _get_session_key()
    nonce = os.urandom(NONCE_LENGTH)
    payload = asdict(session) | {"expiresAt": expires_at}
    plaintext = json.dumps(payload).encode("utf-8")
    ciphertext = AESGCM(key).encrypt(nonce, plaintext, None)
    return base64.urlsafe_b64encode(nonce + ciphertext).decode("ascii").rstrip("=")


def open_session(token: str) -> tuple[MythosSession, str] | None:
    key = _get_session_key()
    try:
        padded = token + "=" * (-len(token) % 4)
        raw = base64.urlsafe_b64decode(padded.encode("ascii"))
        nonce, ciphertext = raw[:NONCE_LENGTH], raw[NONCE_LENGTH:]
        plaintext = AESGCM(key).decrypt(nonce, ciphertext, None)
        payload = json.loads(plaintext.decode("utf-8"))
        if not isinstance(payload, dict):
            return None
        expires_at = payload.pop("expiresAt", None)
        if not isinstance(expires_at, str):
            return None
        parsed_expiry = datetime.fromisoformat(expires_at.replace("Z", "+00:00"))
        if parsed_expiry.tzinfo is None:
            parsed_expiry = parsed_expiry.replace(tzinfo=timezone.utc)
        if parsed_expiry <= datetime.now(timezone.utc):
            return None
        return MythosSession(**payload), expires_at
    except Exception:
        return None
